# TP NoSQL

Prénom : Mehdi

Nom : Boussedra

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [1]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [3]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://mehdi_mongo:5r4las9CeOAQTKMC@cluster-but-sd.q4e4b.mongodb.net/?retryWrites=true&w=majority&appName=cluster-but-sd'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['movies', 'restaurants']


1. Combien de films sont dans la base de données ?

In [10]:
db.movies.count_documents({})

23530

Il y a 23530 films dans la base de donnée

2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [61]:
c = db.movies.aggregate([
    {"$unwind": "$countries"},
    {"$group": {"_id": "$countries", "nb_moyen" : {"$avg" : "$imdb.votes" }}},
    { "$sort": { "nb_moyen": -1}}
])
pd.DataFrame(list(c))

_id       nb_moyen
0                   Malta  210571.800000
1                 Bahamas  206061.000000
2             New Zealand   66589.269231
3    United Arab Emirates   53495.833333
4                     USA   35016.703663
..                    ...            ...
153                Kosovo     104.500000
154               Moldova     104.000000
155           North Korea      92.666667
156                Brunei      51.000000
157               Vanuatu       6.000000

[158 rows x 2 columns]

Voici le nombre moyen de votes par pays : 

3. Donner la liste des différents genres de films.

In [30]:
film_genre = db.movies.distinct(key="genres")
pd.DataFrame(list(film_genre))


0
0        Action
1     Adventure
2     Animation
3     Biography
4        Comedy
5         Crime
6   Documentary
7         Drama
8        Family
9       Fantasy
10    Film-Noir
11      History
12       Horror
13        Music
14      Musical
15      Mystery
16         News
17      Romance
18       Sci-Fi
19        Short
20        Sport
21    Talk-Show
22     Thriller
23          War
24      Western

Voici la liste des differents genres de film

4. Combien de films ont une note IMDb supérieure à 8 ?

In [59]:
note_sup8 = db.movies.count_documents({"imdb.rating": {"$gt": 8}})
     
print(note_sup8)

1177


1177 films ont une note IMDb supérieure à 8

5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [77]:
nom_film = db.movies.find(
    {},
    {"_id": 0 , "title":1, "released":1, "directors":1, }
)
pd.DataFrame(list(nom_film))

title   released  \
0      Intolerance: Love's Struggle Throughout the Ages 1916-09-05   
1                             The Poor Little Rich Girl 1917-03-05   
2                                               The Kid 1921-02-06   
3                                        Miss Lulu Bett 1921-11-01   
4                                          Safety Last! 1923-04-01   
...                                                 ...        ...   
23525                                            Emelie 2015-04-22   
23526                                    Techo y comida 2015-04-23   
23527                                             Nahid 2015-05-16   
23528                                  Coin Locker Girl 2015-04-29   
23529                                 Tryapichnyy soyuz 2015-06-10   

                            directors  
0                     [D.W. Griffith]  
1                  [Maurice Tourneur]  
2                   [Charles Chaplin]  
3               [William C. de Mille]  
4      [Fred C. Newmeyer, Sam Taylor]  
...                               ...  
23525                [Michael Thelin]  
23526      [Juan Miguel del Castillo]  
23527                [Ida Panahandeh]  
23528                   [Jun-hee Han]  
23529            [Mikhail Mestetskiy]  

[23530 rows x 3 columns]

Voici le titre l'année et les realisateurs de tous les films de la base

6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [71]:
plot_dwgriffith = db.movies.find(
    {"directors" : "D.W. Griffith"},
    {"_id": 0, "title": 1, "plot": 1}
    )
pd.DataFrame(list(plot_dwgriffith))
    

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title  
0  Intolerance: Love's Struggle Throughout the Ages  
1                                 A Corner in Wheat  
2    Broken Blossoms or The Yellow Man and the Girl  
3                             The Birth of a Nation

DW Griffith a realise Inolerence : Love's Struggle Troughout the Ages ...

7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [95]:
top5_acteurs = db.movies.aggregate([
    {"$unwind": "$cast"},
    {"$group": {"_id": "$cast", "nb": {"$sum": 1}}},
    {"$sort": {"nb": -1}},
    {"$limit": 5}
])
pd.DataFrame(list(top5_acteurs))

_id  nb
0      Gèrard Depardieu  68
1        Robert De Niro  60
2         Michael Caine  52
3  Marcello Mastroianni  50
4          Bruce Willis  49

Les 5 acteurs les plus populaire sont Gerard Depardieu , Robert de Niro , Michael Caine ...

8. Quelles sont les 6 années avec le plus de films produits ?

In [97]:
top6_annees = db.movies.aggregate([
    {"$group": {"_id": "$year", "nb_films": {"$sum": 1}}},
    {"$sort": {"nb_films": -1}},
    {"$limit": 6}
])

pd.DataFrame(list(top6_annees))


_id  nb_films
0  2013      1220
1  2014      1147
2  2012      1109
3  2011      1040
4  2009       990
5  2010       970

Les 6 années avec le plus de films produits sont 2013,2014,2012 ...

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [119]:
film_avant_2000 = db.movies.count_documents({
    "year" : {"$lt": 2000},
    "imdb.rating" : {"$gt": 7}
})
print("il y a",film_avant_2000," films réalisés avant 2000 et ayant une note supérieur à 7")
                                             
    

il y a 4747  films réalisés avant 2000 et ayant une note supérieur à 7


10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [123]:
top20_real = db.movies.aggregate([
    {"$unwind": "$directors"},
    {"$group": {"_id" : "$directors", "nb": {"$sum": 1}}},
    {"$sort": {"nb": -1}},
    {"$limit": 20}
])
pd.DataFrame(list(top20_real))

_id  nb
0         Woody Allen  40
1           John Ford  35
2       Takashi Miike  34
3         John Huston  34
4       Werner Herzog  33
5     Martin Scorsese  32
6    Alfred Hitchcock  31
7        Sidney Lumet  30
8        George Cukor  29
9       Michael Apted  29
10   Steven Spielberg  29
11    Mario Monicelli  29
12  Steven Soderbergh  28
13          Spike Lee  28
14      Robert Altman  28
15        Wim Wenders  27
16    Jean-Luc Godard  27
17          Ken Loach  27
18         Johnnie To  27
19     Clint Eastwood  27

Woody Allen a réalisé le plus de films.

11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [152]:
note_moyenne = db.movies.aggregate([
    {"$match": {"year": {"$gt" : 2010}}},
    {"$group": {"_id": None, "average_rating": {"$avg" : "$imdb.rating"}}}
])
pd.DataFrame(list(note_moyenne))

_id  average_rating
0  None        6.544033

La note moyenne des films sortis apres 2010 est de 6.5

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [149]:
films_love = db.movies.find(
    {"title": {"$regex": "love", "$options": "i"}}
)

pd.DataFrame(list(films_love))


_id  \
0    573a1390f29313caabcd5c0f   
1    573a1393f29313caabcdbed0   
2    573a1392f29313caabcda515   
3    573a1392f29313caabcdb7b7   
4    573a1392f29313caabcd9b93   
..                        ...   
438  573a13faf29313caabdebb94   
439  573a13f3f29313caabddfd80   
440  573a13f7f29313caabde5e3d   
441  573a13fbf29313caabdedf30   
442  573a13f4f29313caabde1dfd   

                                                  plot  \
0    The story of a poor young woman, separated by ...   
1    A dashing pilot and a vivacious reporter have ...   
2    Mary Barrett is an aspiring Opera singer who i...   
3    Andy's girlfriend Polly is planning to spend C...   
4    A Parisian tailor finds himself posing as a ba...   
..                                                 ...   
438  A group of women in Morocco make a living as p...   
439                                                NaN   
440  Larry Kramer in Love and Anger is a feature-le...   
441  Guy's tragic death is a shock for the Leblanc ...   
442  A couple who have lived together for 76 years ...   

                                genres      rated  \
0                     [Drama, History]  NOT RATED   
1             [Comedy, Drama, Romance]     PASSED   
2                     [Music, Romance]     PASSED   
3                    [Comedy, Romance]     PASSED   
4           [Comedy, Musical, Romance]     PASSED   
..                                 ...        ...   
438                            [Drama]        NaN   
439           [Comedy, Drama, Fantasy]        NaN   
440  [Documentary, Biography, History]        NaN   
441                            [Drama]        NaN   
442     [Documentary, Family, Romance]        NaN   

                                                  cast  num_mflix_comments  \
0    [Lillian Gish, Spottiswoode Aitken, Mary Alden...                 1.0   
1    [Claudette Colbert, Ray Milland, Dennis O'Keef...                 3.0   
2    [Grace Moore, Tullio Carminati, Lyle Talbot, M...                 NaN   
3    [Mickey Rooney, Lewis Stone, Fay Holden, Cecil...                 1.0   
4    [Maurice Chevalier, Jeanette MacDonald, Charle...                 1.0   
..                                                 ...                 ...   
438  [Loubna Abidar, Danny Boushebel, Abdellah Dida...                 NaN   
439  [Vivian Sung, Bruce Lu-Si Bu, Megan Lai, Li-An...                 3.0   
440                                     [Larry Kramer]                 2.0   
441  [Maxim Gaudette, Karelle Tremblay, Valerie Cad...                 1.0   
442                                                NaN                 NaN   

                                                title  \
0    Intolerance: Love's Struggle Throughout the Ages   
1                                      Arise, My Love   
2                                   One Night of Love   
3                               Love Finds Andy Hardy   
4                                     Love Me Tonight   
..                                                ...   
438                                        Much Loved   
439                               Cafè. Waiting. Love   
440                    Larry Kramer in Love and Anger   
441                                    Our Loved Ones   
442                   My Love, Don't Cross That River   

                                              fullplot          countries  \
0    Intolerance and its terrible effects are exami...              [USA]   
1    In 1939, American Tom Martin, who fought in th...              [USA]   
2    Mary Barrett is an aspiring Opera singer who i...              [USA]   
3    Andy's girlfriend Polly is planning to spend C...              [USA]   
4    Maurice Courtelin, a Parisian tailor (Maurice ...              [USA]   
..                                                 ...                ...   
438  A group of women in Morocco make a living as p...  [France, Morocco]   
439                                                NaN   

Le film Arise My love contient le mot Love


13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [147]:
films_rt = db.movies.find(
    {"tomatoes.rotten": {"$gt": 75}},
    {"title": 1, "_id": 0}
)

pd.DataFrame(list(films_rt))


title
0                   Glitter
1            Batman & Robin
2              The Avengers
3       Eye of the Beholder
4            Fantastic Four
..                      ...
712                 Taken 3
713                Blackhat
714  Magic in the Moonlight
715               Insurgent
716             Hot Pursuit

[717 rows x 1 columns]

Glitter a un score Rotten Tomatoed supérieur à 75
